### PR TITLE
IAM Support

### PR DESCRIPTION
**Pull request by sid41**
Implements #13
- Uses fake broker for gp instance testing
- Runs IAM unit tests optionally
- Avoids MT for testing translation APIs

### CODE REVIEW

File: gpclient/gpserviceaccount.py [hunk 1]
```diff
@@ -52,56 +73,95 @@ class GPServiceAccount():
...
+                        create GPServiceAccount""")
+                elif apiKey:
+                    self.__setcreds__(url, instanceId, apiKey)
+                    credentialsSet = True
+                    logging.info("""using user defined environment variables to
+                        create GPServiceAccount supporting IAM auth.""")
         # make sure that all the vars are set
-        assert self.__url, ('url is not a string: <%s>')
+        assert self.__url, ('url is not a string: <%s>' % self.__url)
         assert self.__instanceId, ('instanceId is not a string: <%s>')
```
Comment (srl295) [inline gpclient/gpserviceaccount.py]: same here, needs `% self.__instanceId` ? or is that implied?

File: gpclient/gpserviceaccount.py [hunk 2]
```diff
@@ -52,56 +73,95 @@ class GPServiceAccount():
...
-        assert self.__userId, ('userId is not a string: <%s>')
-        assert self.__password, ('password is not a string: <%s>')
-
-        logging.info(('created GPServiceAccount using url <%s>, ' + \
+        if self.__iamEnabled:
+            assert self.__apiKey, ('apiKey is not a string: <%s>')
```
Comment (srl295) [inline gpclient/gpserviceaccount.py]: same here, needs `% self.__apiKey`  ? or is that implied? (also later cases)

**Review by srl295 — APPROVED** [review]
LGTM with a question about asserts

**Comment by srl295** [comment]
https://travis-ci.org/IBM-Cloud/gp-python-client/jobs/525127742#L691 Why does it fail only on python 3.6? Some odd errs at the top

**Comment by sid41** [comment]
> https://travis-ci.org/IBM-Cloud/gp-python-client/jobs/525127742#L691 Why does it fail only on python 3.6? Some odd errs at the top

I think last build failed because of some transient error (difference between the failing build and the one previous to it was just the line termination chars). I have added some logging in common.py to help detect any errors in init in future. The error on the top is still there. It was there in previous successful builds too. Basically dateutil lib which is loaded is not compliant with py3 print command...

File: gpclient/gpserviceaccount.py (hunk 1 above)
Comment (sid41) [reply gpclient/gpserviceaccount.py]: No string will ever evaluate to false in here. I think original author's intent was to inform if this assert fails, say this should be string type, not really familiar with the conventions in py. Does seem misleading - agree.

Only following values of variable will lead assert to fail:
None
False
zero of any numeric type, for example, 0, 0L, 0.0, 0j.
any empty sequence, for example, '', (), [].
any empty mapping, for example, {}.
instances of user-defined classes, if the class defines a __nonzero__() or __len__() method, when that method returns the integer zero or bool value False.
src: https://docs.python.org/2.4/lib/truth.html

Comment (sid41) [reply gpclient/gpserviceaccount.py]: hmm ok, tried something in py console, should do %... for all to print out the invalid values

**Comment by sid41** [comment]
> > https://travis-ci.org/IBM-Cloud/gp-python-client/jobs/525127742#L691 Why does it fail only on python 3.6? Some odd errs at the top
> 
> I think last build failed because of some transient error (difference between the failing build and the one previous to it was just the line termination chars). I have added some logging in common.py to help detect any errors in init in future. The error on the top is still there. It was there in previous successful builds too. Basically dateutil lib which is loaded is not compliant with py3 print command...

I did a text search in project for "dateutils"  and don't see any dependency in project- probably can be removed.

**Pull request merged**